MAINT: stats: Rewrite pearsonr.

WarrenWeckesser opened the pull request:
The calculation of r in pearsonr(x, y) is rewritten in a more
robust form.

Let

    xm = x - mean(x)
    ym = y - mean(y)

In this commit, the calculation of r is changed from (roughly)

    r = dot(xm, ym)/sqrt((xm**2).sum() * (ym**2).sum())

to

    r = dot(xm/norm(xm), ym/norm(ym))

where the norm is calculated using `scipy.linalg.norm`.

`scipy.linalg.norm` avoids the underflow and overflow of intermediate results
that was the source of a couple bug reports.

The function has the following additional argument validation:

 * A ValueError is raised if the lengths of x and y are not the same.
 * A ValueError is raised if the length is less than 2.

New warnings are generated in the following cases:

 * If an input is *constant*, a `PearsonRConstantInputWarning` is generated,
   and nan is returned for both r and prob.  The correlation coefficient
   is not defined in this case.
 * If an input is *nearly* constant, a `PearsonRNearConstantInputWarning` is generated,
   but the calculation proceeds.  The condition for x being classified as
   near constant is

       norm(xm)/abs(mean(x)) < 1e-13

   (This is a heuristic condition, and it may generate false warnings.)

Another change is that inputs with length 2 are handled as a special case.
The only possible values of r in this case are +1 or -1, and in both cases,
the p-value is 1.

The calculation of the p-value has also been rewritten. It now has the simpler
form

    ab = len(x)/2 - 1
    prob = 2*special.btdtr(ab, ab, 0.5*(1 - abs(r)))

`special.btdtr` is the CDF of the beta distribution.

There were two existing tests that did not pass with the new code.  For the
inputs x = [0.0, 1.0, 2.0] with y = x and y = -x, the previous version of
pearsonr returned *exactly* 1 and -1, respectively, with p-value 0 in both
cases.  With the new code, we have:

    In [14]: x = np.arange(3.0)

    In [15]: pearsonr(x, x)
    Out[15]: (0.9999999999999998, 1.3415758552508151e-08)

    In [16]: pearsonr(x, -x)
    Out[16]: (-0.9999999999999998, 1.3415758552508151e-08)

The r values differ from the exact values by 2 ULPs.  The reason the p-values
are then so large (well, large compared to the exact value 0) is that, for n=3,
the p-value for r near 1 or -1 grows like the square root of 1 - abs(r), and
np.spacing(1) is about 2.22e-16.  When n is larger, the p-value is not so
sensitive to small perturbations in r.

New tests have been added to verify that problems reported with the old
code are now fixed.

Closes gh-3728.
Closes gh-7730.
Closes gh-8980.
Closes gh-9353.
Closes gh-9406.

WarrenWeckesser commented:
I didn't touch `mstats.pearsonr` in this pull request.  If these changes look good, `mstats.pearsonr` can be given a similar update in a separate pull request.

WarrenWeckesser commented:
In the new tests, I commented that the nontrivial expected values were computed using mpmath.  In case anyone is interested, I put the code for this (and many other mpmath calculations) in a repository at https://github.com/WarrenWeckesser/mpsci.  For example,

```
In [52]: import mpmath

In [53]: mpmath.mp.dps = 80

In [54]: from mpsci import stats

In [55]: x = [0.004434375, 0.004756007, 0.003911996, 0.0038005, 0.003409971]

In [56]: y = [2.48e-188, 7.41e-181, 4.09e-208, 2.08e-223, 2.66e-245]

In [57]: r, p = stats.pearsonr(x, y)

In [58]: r
Out[58]: mpf('0.72729305407504503892383541873085717035420594019716673586636186070999911443256203245')

In [59]: p
Out[59]: mpf('0.16378054295332022029263645440491518881019718162012651605141086778254992932854963897')
```

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -3029,28 +3057,51 @@ def pearsonr(x, y):
-        )
+
+    halfdf = n/2 - 1
+    prob = 2*special.btdtr(halfdf, halfdf, 0.5*(1 - abs(r)))
```
how does this relate to the t-distribution before?
https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#Testing_using_Student's_t-distribution

WarrenWeckesser replied:
For the null hypothesis that the population correlation ρ is 0, the distribution of r is the beta distribution on [-1, 1] with equal shape parameters a = b = n/2 - 1.  (This is the result that "Student" reported in 1908; there is no mention of the t distribution in that paper.)  The distribution is referred to as the exact distribution in the wikipedia article that you linked to (look in the subsection titled "Using the exact distribution", immediately after the subsection that you linked to).  The code in this pull request uses the exact distribution.

The t-distribution comes from the fact that t = r*sqrt((n-2)/(1-r**2)) is t-distributed under the same null hypothesis.  There is a connection between the t-distribution and the beta distribution, and the old code used that connection to compute the p-value.  The new code is mathematically equivalent to the old code.

(Note that the old code is needlessly complicated: it computes t_squared, and then passes df/(df+t_squared) to `betainc`.  But if you plug the expression for t_squared into df/(df+t_squared), it simplifies to 1-r**2!)

I didn't touch the docstring in this pull request, but I will.  I realize this might make https://github.com/scipy/scipy/pull/9498 obsolete.  (@rlucas7, sorry I keep stepping on your pull requests!)

In the docstring, I can say that the p-value is computed with the exact distribution of r, and give a brief explanation of what that distribution is.  I'll also add comments in the code to explain exactly where `2*special.btdtr(halfdf, halfdf, 0.5*(1 - abs(r)))` comes from.

rlucas7 replied:
>I didn't touch the docstring in this pull request, but I will. I realize this might make #9498 >obsolete. (@rlucas7, sorry I keep stepping on your pull requests!)

@WarrenWeckesser no worries, I'm glad the method is fixed, and I prefer your solution which addresses all of the issues in 1 PR. I'm all about better code, not getting more PRs in :).

>n the docstring, I can say that the p-value is computed with the exact distribution of r, and >give a brief explanation of what that distribution is.

Here I would recommend to use something other than the word "exact" because "exact" usually means a small sample permutation test where the code generates all possible permutations and calculates a test statistic (or p-value) from the-empirically generated-distribution.

Also, this is only the correct distribution if the observations come from normal distributions. The observations may not come from normals and then this would be an asymptotic approximation (assuming the distribution that generates the data has finite mean and variance so CLT can kick in).

WarrenWeckesser replied:
@rlucas7, take a look at the new Notes section of the docstring, where, after giving the formula for the PDF f(r), I say "This is sometimes referred to as the exact distribution of r."  The distribution is "exact" in the sense that it is not an asymptotic result.  I know that "exact" is also used in other contexts (e.g. Fisher's exact test), and the overloading of the meaning is unfortunate.  I'm sure I have seen this use of "exact" in reference to this distribution for Pearson's r more than once, but at the moment the only reference that I can find is the wikipedia page: https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#Using_the_exact_distribution

Regarding the normal assumption: also in the Notes section, before giving the formula for the PDF, I say "Under the assumption that x and y are drawn from independent normal distributions (so the population correlation coefficient is 0) ..."

tylerjereddy replied:
Sounds like @rlucas7 is +1 on the PR below & the docs have the appropriate content now, so marking this as resolved

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -3029,28 +3057,51 @@ def pearsonr(x, y):
 
     # Presumably, if abs(r) > 1, then it is only some small artifact of
     # floating point arithmetic.
     r = max(min(r, 1.0), -1.0)
```
in one of the issues, this line caused a problem when r = inf or -inf

WarrenWeckesser replied:
That's https://github.com/scipy/scipy/issues/9353.  In that issue, the values in `y` are so small that in the calculation `sum(y**2)`, the squared terms underflow to 0, which caused the denominator in the calculation of `r` to be 0.  That problem is fixed by the reformulation in this pull request, and the data from that issue is used in one of the unit tests.

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -3029,28 +3057,51 @@ def pearsonr(x, y):
+        return np.nan, np.nan
+
+    if n == 2:
+        return np.float64(np.sign(x[1] - x[0])*np.sign(y[1] - y[0])), 1.0
```
This probably should be explained in the documentation as a special case, otherwise user will ask why the p-value is one

WarrenWeckesser replied:
Agreed.  I'll add an explanation to the docstring.

WarrenWeckesser replied:
I added an explanation of this to the "Notes" section of the docstring.

tylerjereddy replied:
The `n == 2` case is now explained in the docstring, so marking this as resolved

WarrenWeckesser commented:
@chrisb83, thanks for the comments.  I'll work on updating the docstring and code comments.

rlucas7 commented:
@WarrenWeckesser In the condition to generate warning:

>If an input is nearly constant, a PearsonRNearConstantWarning is generated,
>but the calculation proceeds. The condition for x being classified as
>near constant is

>norm(xm)/abs(mean(x)) < 1e-13

I've seen a couple instances where depending on the word size of the machine the error tolerance might not hold. I'm no expert on hardware (nor on the commonly used hardware of users of SciPy) so I'll just pose the question: is this a high enough tolerance to not generate warnings in most common systems? I'm not sure if this use of the `mpmath` obviates the concern or not.

WarrenWeckesser commented:
@rlucas7 wrote
> is this a high enough tolerance to not generate warnings in most common systems? 

In my most recent update, I try to ensure that the calculations are done with the precision of 64 bit floating point (or higher, if the inputs are `np.longdouble` on a system where `np.longdouble` actually provides higher precision than 64 bits).

The test for "nearly constant" is a heuristic, and I can't give a rigorous justification for the threshold value 1e-13.  The test is not guaranteed to detect bad results, and it can also generate false positives.  Actually, even the notion of what should be considered a "false positive" is fuzzy.  To how many significant digits should r be correct before we call it "good enough"?  And even if we had the answer to that question, I don't know of a reasonably efficient method to ensure that r has been computed that accurately.

The reason the warning is there is to alert the user to potentially *really bad* results.  For example, here's an extreme case:

```
In [28]: from scipy.stats import pearsonr

In [29]: x = np.arange(10) + 1e15

In [30]: y = -0.5*np.sin(np.linspace(0, 2, len(x))) + 2e15

In [31]: pearsonr(x, y)
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0+ec379c8-py3.6-macosx-10.7-x86_64.egg/scipy/stats/stats.py:3156: PearsonRNearConstantInputWarning: An input array is nearly constant; the computed correlation coefficent may be inaccurate.
  warnings.warn(PearsonRNearConstantInputWarning())
Out[31]: (-0.6695652173913044, 0.03418930694603446)
```
`pearsonr` computed `r = -0.6695652173913044`.  Now use the `mpmath`-based code to compute a more accurate value:

```
In [32]: from mpsci import stats

In [33]: import mpmath

In [34]: mpmath.mp.dps = 80

In [35]: stats.pearsonr(x, y)
Out[35]: 
(mpf('-0.8703882797784891908864562830634741391242824936337564193484662834190758668874855475'),
 mpf('0.0010528257933665392739547153364779790138010560167993011095113153680111269784806330852'))
```
The actual value is `r = -0.870388279778489`.  In this case the warning is helpful.

If we make the example slightly less extreme by lowering the exponent of those constant offsets from 15 to 13...
```
In [38]: x = np.arange(10) + 1e13

In [39]: y = -0.5*np.sin(np.linspace(0, 2, len(x))) + 2e13

In [40]: pearsonr(x, y)
/Users/warren/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0+ec379c8-py3.6-macosx-10.7-x86_64.egg/scipy/stats/stats.py:3156: PearsonRNearConstantInputWarning: An input array is nearly constant; the computed correlation coefficent may be inaccurate.
  warnings.warn(PearsonRNearConstantInputWarning())
Out[40]: (-0.9084102476236772, 0.000275306006401631)
```
we still get a warning.  In fact, the r value is accurate to 3 significant digits:
```
In [41]: stats.pearsonr(x, y)
Out[41]: 
(mpf('-0.90861016484050085312613384635342519571369356334706754822950384796382168010704943808'),
 mpf('0.0002729779145539270031014645957952420149808094617731500756779293175987705031464615037'))
```
So in that case, a user who is satisfied with 3 significant digits might be annoyed at the spurious warning.

And in case you're wondering: if the exponent is lowered to 12--which gives arrays that are still darn close to constant--no warning is generated, and the r value computed by `pearsonr` agrees with the `mpmath`-based calculation to 6 digits:

```
In [46]: x = np.arange(10) + 1e12

In [47]: y = -0.5*np.sin(np.linspace(0, 2, len(x))) + 2e12

In [48]: pearsonr(x, y)
Out[48]: (-0.9092112104610951, 0.00026606567693838426)

In [49]: stats.pearsonr(x, y)
Out[49]: 
(mpf('-0.90921182685311774154350338076254087062170180168583372833257087142909862987252274978'),
 mpf('0.00026605865496040174239800491438852243123083135034930679827842613771743090051875876115'))
```

WarrenWeckesser commented:
I pushed an update with better docs.  I don't know if the pypy3 failure is related to this pull request; after 93% of the tests finished, the log says "Too long with no output (exceeded 10m0s)".

tylerjereddy commented:
>  "Too long with no output (exceeded 10m0s)"

Pretty sure that's been happening now and then -- I think there's a way to add `no_output_timeout` directive if we haven't yet.

rlucas7 commented:
> @rlucas7 wrote
> 
> > is this a high enough tolerance to not generate warnings in most common systems?
> 
> In my most recent update, I try to ensure that the calculations are done with the precision of 64 bit floating point (or higher, if the inputs are `np.longdouble` on a system where `np.longdouble` actually provides higher precision than 64 bits).
> 
> The test for "nearly constant" is a heuristic, and I can't give a rigorous justification for the threshold value 1e-13. The test is not guaranteed to detect bad results, and it can also generate false positives. Actually, even the notion of what should be considered a "false positive" is fuzzy. To how many significant digits should r be correct before we call it "good enough"? And even if we had the answer to that question, I don't know of a reasonably efficient method to ensure that r has been computed that accurately.
> 
> The reason the warning is there is to alert the user to potentially _really bad_ results. For example, here's an extreme case:

Thanks @WarrenWeckesser I appreciate the thorough response. This PR LGTM 👍.

chrisb83 commented:
@WarrenWeckesser very nice improvement of the current implementation. Testing looks very comprehensive (as a sanity check, I reproduced the tests `test_very_large_input_values` and `test_very_small_input_values` in R and get the same results). Documentation looks good as well.

One suggestion as in #9498: 
Say that if data is non-normal, the p-value is only an approximation, see [2]_ for a discussion on the reliability of the p-value in the non-normal case. (it took me a while to find that reference, would be nice to add it)

2 = C.J. Kowalski: On the Effects of Non-Normality on the Distribution of the Sample Product-Moment Correlation Coefficient.Journal of the Royal Statistical Society. Series C (Applied Statistics), Vol. 21, No. 1 (1972), pp. 1-12

One question: why do you propose to create two specific warning classes instead of just `warnings.warn(…)` that are part of the public API. I was just wondering that if we do this is other functions as well, we have lots of warning types.

chrisb83 commented:
@WarrenWeckesser Any update on this PR? It would be great to have your improvements in 1.3.0

WarrenWeckesser commented:
@chrisb83, great timing!  I just starting getting back into this yesterday.  A belated "thanks" for your comments back in December.

WarrenWeckesser commented:
@chrisb83 wrote:

> One question: why do you propose to create two specific warning classes instead of just warnings.warn(…) that are part of the public API. I was just wondering that if we do this is other functions as well, we have lots of warning types.

The advantage of creating specific warnings for specific conditions is that it gives the user a reliable way to control how the warnings are handled.  If a user wants to filter the warnings, and we only use, say

    warnings.warn("An input array is nearly constant; the computed "
                  "correlation coefficent may be inaccurate.")

(which generates a `UserWarning`), or

    warnings.warn(RuntimeWarning("An input array is nearly constant; the computed "
                                 "correlation coefficent may be inaccurate."))

the user's code doesn't necessarily know which condition generated the warning unless the message string is inspected. That is, if all the warnings are `RuntimeWarnings`, and a user wants to ignore only the warning generated when an input is nearly constant, the filter would have to be set to match a `RuntimeWarning` along with some pattern within the warning's message.  E.g.

    warnings.filterwarnings('ignore', RuntimeWarning, message='.*nearly constant')

I don't think we have a policy of ensuring that the message in a warning doesn't change between releases, so that approach is fragile.

If instead we create a specific warning, the user can filter it without worrying about future changes to the warning message:

    warnings.filterwarnings('ignore', category=PearsonRNearConstantInputWarning)

There are quite a few other places in the scipy code where specific warnings are defined; a search for something like `class .*arning` among all the python files should find most of them.

I haven't checked if these are all also included in a "Warns" section of the appropriate docstrings, but they probably should be. :smile:

WarrenWeckesser commented:
@chrisb83, I'm adding a reference to the Kowalski paper to the docstring.

WarrenWeckesser commented:
The failures on TravisCI are https://github.com/scipy/scipy/issues/9308.

WarrenWeckesser commented:
I added this to the 1.3 milestone, but the release manager has the final decision.

rgommers commented:
@WarrenWeckesser is it ready for final review? Would be great to get this in.

WarrenWeckesser commented:
@rgommers, when I finish the review of https://github.com/scipy/scipy/pull/8681 that I'm working on, I'll take a fresh look at this PR.  If I'm still happy with it, I'll request a final review.

WarrenWeckesser commented:
I made a few more small tweaks to the code and squashed the commits.  This is ready for final review.  I think I've addressed all the previous comments, but if not, let me know.

chrisb83 commented:
@WarrenWeckesser thanks for adding the reference to the article of Kowalski.

I agree that having specific warnings is a good idea. Two more questions:

- should we replace PearsonRConstantInputWarning by ConstantInputWarning, I think this could be a generic warning whereas PearsonRNearConstantInputWarning might be indeed a specific warning for PearsonR
- do we want to show the warnings in the overview generated by __init__.py? I fear this might become a long list that distracts attention from the main stats functions and pointing out the warnings in the section `Warns` of the function is enough. Just my opinion :)

Let me read through the docstring one more time tomorrow. I think this should really go into 1.3.0 .

WarrenWeckesser commented:
@chrisb83 wrote

> should we replace PearsonRConstantInputWarning by ConstantInputWarning, I think this could be a generic warning whereas PearsonRNearConstantInputWarning might be indeed a specific warning for PearsonR

There are probably other functions that could use a similar pattern, but currently, it *is* specific to `pearsonr`.  I'd rather not try to generalize this without concrete use-cases.   And if/when we implement a similar behavior for other functions, I don't think there would be anything wrong with another function, say `foo()`, having its own `FooConstantInputWarning`. 

> do we want to show the warnings in the overview generated by init.py? 

I debated this, too.  I figured they are part of the public API of `scipy.stats`, so they should be there.  But I don't mind taking them out.  They are still documented in the `pearsonr` docstring, and maybe that's enough.  Any other opinions out there?

larsoner commented:
> I figured they are part of the public API of scipy.stats, so they should be there. But I don't mind taking them out. They are still documented in the pearsonr docstring, and maybe that's enough. Any other opinions out there?

No strong opinion from me. Their inclusion does not seem too obtrusive at the end of the API docs so if forced to choose, I'd keep them:

https://12748-1460385-gh.circle-artifacts.com/0/html-scipyorg/stats.html#warnings-used-in-scipy-stats

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -3231,23 +3234,47 @@ def f_oneway(*args):
+    Pearson correlation coefficient and p-value for testing non-correlation.
 
-    The Pearson correlation coefficient measures the linear relationship
+    The Pearson correlation coefficient [1]_ measures the linear relationship
```
One could replace "Strictly speaking, Pearson's correlation requires that each dataset be normally distributed, and not necessarily zero-mean. " by sth like "The calculation of the p-value relies on the assumption that each dataset is normally distributed."

WarrenWeckesser replied:
:+1: I'll make a change like that.

tylerjereddy replied:
I see this fixed in the diff -- marking as resolved

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -3261,7 +3288,23 @@ def pearsonr(x, y):
+
+    See Also
+    --------
+    spearmanr : Calculate a Spearman rank-order correlation coefficient
```
one could add kendalltau as well

WarrenWeckesser replied:
:+1: Sure.

tylerjereddy replied:
Also, fixed in diff now

WarrenWeckesser commented:
I pushed a new commit with @chrisb83's latest suggested changes to the docstring.

tylerjereddy commented on scipy/stats/tests/test_stats.py:
```diff
@@ -255,19 +255,23 @@ def test_pROUNDROUND(self):
+        # square root of the error in r.
+        assert_allclose(prob, 0.0, atol=np.sqrt(2*np.spacing(1.0)))
+
+    def test_r_almost_exactly_neg1(self):
```
This could be condensed into the positive version of the test above with parametrization, but not worth the effort at this stage in my view.

tylerjereddy commented on scipy/stats/tests/test_stats.py:
```diff
@@ -276,7 +280,96 @@ def test_basic(self):
+        assert_allclose(r, 0.8660254037844386)
+        assert_allclose(p, 0.011724811003954638)
+
+    def _test_extremely_large_input_values(self):
```
Why is there a leading underscore in this test name? Probably a typo.

If you remove that, you can verify locally that `python runtests.py --mode=full -t "scipy/stats/tests/test_stats.py::TestCorrPearsonr"` will execute one additional test.

WarrenWeckesser replied:
I don't know why I put a leading underscore in the name, and I'm embarrassed that I didn't notice that the test wasn't being run 😟 .  I fixed it in my latest update.

tylerjereddy commented on scipy/stats/tests/test_stats.py:
```diff
@@ -276,7 +280,96 @@ def test_basic(self):
+        assert_equal(r, -1)
+        assert_equal(p, 1)
+
+    def test_more_basic_examples(self):
```
A bit more value to parametrizing here for separate reports on failure, but again, not likely worth worrying about at this stage.

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -3270,50 +3314,131 @@ def pearsonr(x, y):
+        raise ValueError('x and y must have the same length.')
+
+    if n < 2:
+        raise ValueError('x and y must have length at least 2.')
```
Probably not crucial because of the simplicity, but I ran line coverage locally:

![image](https://user-images.githubusercontent.com/7903078/56327434-35792600-612f-11e9-8478-63671ac4279c.png)

tylerjereddy replied:
Every single other line in `pearsonr` is covered, so that's good.

WarrenWeckesser replied:
I added a couple tests that exercise these conditions.

WarrenWeckesser commented:
Thanks @tylerjereddy. I didn't change the tests to use parametrization (pytest... yet another mini-framework with yet another way to write a loop :), but I fixed the other issues that you pointed out.

tylerjereddy commented:
Thanks @WarrenWeckesser and all the reviewers.

Is there a release note here: https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.3.0

I can't remember off the top of my head if we need to tag the closed issues with the milestone for the scraping code to pick them up when I do release notes workflow, maybe I'll check that now.

WarrenWeckesser commented:
I don't know, I haven't managed a release before.  My default way to handle a question like that is "Ask Ralf". 😄 

Thanks, all.  Nice to close five issues in one swell foop.

chrisb83 commented:
This also closed #8789, so a total of six issues :)

WarrenWeckesser commented:
@chrisb83, thanks, I forgot about that one!

rgommers commented:
Great to get this one in, thanks Warren and all reviewers!

> I can't remember off the top of my head if we need to tag the closed issues with the milestone for the scraping code to pick them up when I do release notes workflow, maybe I'll check that now.

Yes. The script is `tools/gh_lists.py` and it uses the milestones we assign to issues and PRs. NumPy now has a more intelligent way I believe that we could copy, however I think there's value anyway in being consistent with tagging PRs and issues with milestones when they were merged/solved. When I find an issue, I'd rather see right there in what release it was solved than having to grep `doc/release/`.

rlucas7 commented:
Yeah I’m +1 just seeing now. Will be good to have in 1.3.0! 

Sincerely, 

-Lucas Roberts 

> On Apr 17, 2019, at 6:46 PM, Tyler Reddy <notifications@github.com> wrote:
> 
> @tylerjereddy commented on this pull request.
> 
> In scipy/stats/stats.py:
> 
> >  
> -    The Pearson correlation coefficient measures the linear relationship
> +    The Pearson correlation coefficient [1]_ measures the linear relationship
> I see this fixed in the diff -- marking as resolved
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

tylerjereddy commented:
I added a draft release note for this PR--please do fix it up if there's anything off about it: 

https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.3.0#scipystats-improvements

rgommers commented:
What is there looks good. I didn't go through merged PRs.

bradknox commented:
I'm not following this conversation closely and could be missing something, but the problematic sentence discussed in [#8789](https://github.com/scipy/scipy/issues/8789) is still in master and [the online docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.spearmanr.html), despite it being considered resolved above.
